cep-recipe-jinja.md

```diff
@@ -0,0 +1,511 @@
+pin expressions.
+
+- `lower_bound='x.x', upper_bound='x.x'` would result in `>=1.21,<1.22.0a0`
+- `lower_bound='x.x.x', max_pin='x'` would result in `>=1.21.3,<2.0a0`
+- `lower_bound=None, upper_bound='x'` would result in `<2.0a0`
+- `lower_bound='x.x.x.x', upper_bound=None` would result in `>=1.21.3`
+- `exact=True` would result in `==1.21.3=h123456_5`
+
+The function should error if `exact` is `True` and `min_pin` or `max_pin` are
+set.
+
+Given the following version `1.2.3`, we get the following results:
+
+- default values: `lower_bound='x.x.x.x.x.x', upper_bound='x'` -> `>=1.2.3,<2.0a0`
+- `lower_bound='1.0', upper_bound='x.x'` -> `>1.0,<1.3.0a0`
+- `lower_bound='x.x', upper_bound='2.0'` -> `>1.2,<2.0`
+- `lower_bound=None, upper_bound='x'` -> `<2.0a0`
+- `lower_bound='x.x.x.x', upper_bound=None` -> `>=1.2.3`
+
+For an input of the form: `9e` (jpeg style version)
+
+- `lower_bound='x', upper_bound='x'` -> `>=9e,<10a`
+
+For an input of the form: `1.1.1j` (openssl style version)
+
+- `lower_bound='x.x.x', upper_bound='x'` -> `>=1.1.1j,<2.0a0`
+- `lower_bound='x.x.x', upper_bound='x.x'` -> `>=1.1.1j,<1.2.0a0`
+- `lower_bound='x.x.x', upper_bound='x.x.x'` -> `>=1.1.1j,<1.1.2a`
+
+### The `pin_compatible` function
+
+Pin compatible will pin the dependency to the same version as "previously"
+resolved in the `host` or `build` environment. This is useful to ensure that the
+same package is used at run time as was used at build time.
+
+Example:
+
+```yaml
+requirements:
+  host:
+    - numpy
+  run:
+    - ${{ pin_compatible('numpy', exact=True) }}
+    # or alternatives
+    # - ${{ pin_compatible('numpy', lower_bound='x.x.x', upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound=None, upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound="1.0", upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound="1.0", upper_bound="2.0") }}
+```
+
+### The `pin_subpackage` function
+
+Pin subpackage will pin the dependency to the same version as another
+sub-package from the recipe (or the current package itself). This is useful to
+ensure that multiple outputs from a recipe are linked together or to export the
+correct `run_exports` for a package.
+
+Example:
+
+```yaml
+outputs:
+  - package:
+      name: libfoo
+      version: "1.2.3"
+  - package:
+      name: foo
+      version: "1.2.3"
+    requirements:
+      run:
+        - ${{ pin_subpackage('libfoo', exact=True) }}
+```
+
+## The `match` function
+
+The `match` function is used to match a variant with a version spec. It returns
+`true` if the version spec matches the variant and `false` otherwise.
+
+For example, it can be used in the following way:
+
+```yaml
+requirements:
+  - $ {{ "six" if match(python, "<3.8") }}
+  - $ {{ "six" if match(python, "3.8") }}
+  - $ {{ "six" if match(python, "==3.8") }}
+  - $ {{ "six" if match(python, "3.8.*") }}
+  - $ {{ "six" if match(python, ">=3.8,<3.10") }}
```

Comment:
```suggestion
  - ${{ "six" if match(python, "<3.8") }}
  - ${{ "six" if match(python, "3.8") }}
  - ${{ "six" if match(python, "==3.8") }}
  - ${{ "six" if match(python, "3.8.*") }}
  - ${{ "six" if match(python, ">=3.8,<3.10") }}
```
(BTW, thanks for going with the string-based Conda version spec format here!)

Reply:
Thanks!